### PR TITLE
Validate and log data during transform

### PR DESCRIPTION
**Pull request by pshapiro4broad**
This is one way to validate (and log validations) while transforming input. I used the GDC transform code because the PDC wasn't merged up yet. It would be possible to pull out the validation code to the top level transform, and validate all DC data in one place, as long as there's no DC specific check you'd want to do on the data.

This should address https://github.com/CancerDataAggregator/transform/issues/8

### CODE REVIEW

File: cdatransform/transform/main.py [hunk 1]
```diff
@@ -30,8 +26,8 @@ def filter_cases(reader, case_list):
             break
         else:
             if case.get("id") in cases:
-                cases.pop(case.get("id"))
-                yield case
+                cases.remove(case.get("id"))
+            yield case
```
Comment (kghose) [inline cdatransform/transform/main.py:30]: @pshapiro4broad there is a bug introduced here. We should eventually have a test to cover this.

The yield should be executed under the `if`. The indent has been lost in line 30.  Without the indent we will get the case returned each loop. Welcome to Python!

Comment (pshapiro4broad) [reply cdatransform/transform/main.py:30]: I'll have to check my IDE settings, this change didn't show up in my local diff. Most likely I have a setting enabled to hide whitespace only changes.

And yes having a test would've been good here!

File: cdatransform/transform/gdclib.py [hunk 2]
```diff
@@ -26,31 +28,42 @@ def research_subject(tip, orig, **kwargs):
...
+def entity_to_specimen(transform_in_progress, original, log: LogValidation, **kwargs):
     """Convert samples, portions and aliquots to specimens"""
     transform_in_progress["Specimen"] = [
         specimen_from_entity(*s)
         for s in get_entities(original)
     ]
+    for specimen in transform_in_progress["Specimen"]:
+        for field in ["primary_disease_type", "source_material_type", "anatomical_site"]:
+            log.distinct(specimen, field)
+        # days to birth is negative days from birth until diagnosis. 7300 days is 200 years.
```
Comment (kghose) [inline cdatransform/transform/gdclib.py:64]: Small typo here.

File: cdatransform/transform/validate.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
+from collections import defaultdict
+from typing import Dict, Set
+
+
+class LogValidation:
+
+    def __init__(self) -> None:
+        # map of field name -> set of values
+        self._distinct_fields: Dict[str, Set] = defaultdict(set)
```
Comment (kghose) [inline cdatransform/transform/validate.py:9]: This is neat!

**Review by kghose — APPROVED** [review]
@pshapiro4broad This looks great! There is only one typo in a comment. If you have time to fix that, otherwise we can merge.

**Review by kghose — APPROVED** [review]
Sold!

**Pull request merged**